### PR TITLE
Implement dataset source and hash tracking.

### PR DESCRIPTION
**Pull request by jmchilton**
Generalization of #4659, dataset checksum should also indicate if they are the checksum for the top-level data or data within extra files path. These models get closer to accounting for that. Extend this work to also allow tracking the source a dataset (or part of a dataset) was generated from as well as the hash of those objects. Includes initial support in the Upload 2.0 API to track the source of certain URLs coming in, the hash of certain things coming in, and an option to validate against those.

This can be the basis of doing a lot of cool stuff in the rule based uploader, dealing with remote data (dev/users customizing these things as part of model store imports - see #7367 for instance), cloud stuff, etc.. not sure what a minimum viable product people would want to see is.

### CODE REVIEW

File: lib/galaxy/util/hash_util.py [hunk 1]
```diff
@@ -17,11 +17,25 @@
...
+    "SHA-256": sha256,
+    "SHA-512": sha512,
+}
+HASH_NAMES = list(HASH_NAME_MAP.keys())
+
+
+def memory_bound_hexdigest(hash_func=None, hash_func_name=None, path=None, file=None):
+    if hash_func is None:
+        assert hash_func_name is not None
+        assert hash_func_name in HASH_NAME_MAP
```
Comment (nsoranzo) [inline lib/galaxy/util/hash_util.py]: This assert is not needed.

File: test/api/test_libraries.py [hunk 2]
```diff
@@ -135,19 +135,42 @@ def test_fetch_zip_to_folder(self):
 
     def test_fetch_single_url_to_folder(self):
         history_id, library, destination = self._setup_fetch_to_folder("single_url")
-        items = [{"src": "url", "url": "https://raw.githubusercontent.com/galaxyproject/galaxy/dev/test-data/4.bed"}]
+        items = [{"src": "url", "url": "https://raw.githubusercontent.com/galaxyproject/galaxy/dev/test-data/4.bed", "MD5": "37b59762b59fff860460522d271bc111"}]
```
Comment (nsoranzo) [inline test/api/test_libraries.py]: Split in multiple lines?

File: test/api/test_libraries.py [hunk 3]
```diff
@@ -135,19 +135,42 @@ def test_fetch_zip_to_folder(self):
...
         }
         self.dataset_populator.fetch(payload)
         dataset = self.library_populator.get_library_contents_with_path(library["id"], "/4.bed")
         assert dataset["file_size"] == 61, dataset
 
+    def test_fetch_failed_validation(self):
+        # Exception handling is really rough here - we should be creating a dataset in error instead
+        # of just failing the job like this.
+        history_id, library, destination = self._setup_fetch_to_folder("single_url")
+        items = [{"src": "url", "url": "https://raw.githubusercontent.com/galaxyproject/galaxy/dev/test-data/4.bed", "MD5": "37b59762b59fff860460522d271bc112"}]
```
Comment (nsoranzo) [inline test/api/test_libraries.py]: Split in multiple lines?

File: lib/galaxy/model/migrate/versions/0148_add_checksum_table.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        log.exception("Adding dataset source and hash tables failed.")
+
+
+def downgrade(migrate_engine):
+    metadata.bind = migrate_engine
+    metadata.reflect()
+    try:
+        dataset_source_table.create()
+        dataset_hash_table.create()
+        dataset_source_hash_table.create()
```
Comment (nsoranzo) [inline lib/galaxy/model/migrate/versions/0148_add_checksum_table.py]: Tables should be dropped here, not created.

File: lib/galaxy/model/mapping.py [hunk 5]
```diff
@@ -223,6 +223,32 @@
...
+    Column("dataset_id", Integer, ForeignKey("dataset.id"), index=True),
+    Column("hash_function", TEXT),
+    Column("hash_value", TEXT),
+    Column("extra_files_path", TEXT),
+)
+
+model.DatasetSourceHash.table = Table(
+    "dataset_source_hash", metadata,
+    Column("id", Integer, primary_key=True),
+    Column("dataset_id", Integer, ForeignKey("dataset_source.id"), index=True),
```
Comment (nsoranzo) [inline lib/galaxy/model/mapping.py]: Should this column be called `dataset_source_id` to avoid confusion (also below)?

Comment (jmchilton) [reply lib/galaxy/model/mapping.py]: Definitely - that is a copy and paste problem - sorry about that.

**Comment by martenson** [comment]
One of the cool things we could do is to use hashes to deduplicate our storage (i.e. store every dataset only once).

**Comment by martenson** [comment]
xref https://github.com/galaxyproject/galaxy/issues/7520

**Pull request merged**